File: samples/client/3_1_0_unit_test/java/src/test/java/unit_test_api/components/schemas/NestedItemsTest.java
package unit_test_api.components.schemas;

import org.junit.Assert;
import org.junit.Test;
import unit_test_api.configurations.JsonSchemaKeywordFlags;
import unit_test_api.configurations.SchemaConfiguration;
import unit_test_api.exceptions.ValidationException;
import unit_test_api.schemas.validation.MapUtils;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.AbstractMap;

public class NestedItemsTest {
    static final SchemaConfiguration configuration = new SchemaConfiguration(new JsonSchemaKeywordFlags.Builder().format().build());

    @Test
    public void testNestedArrayWithInvalidTypeFails() {
        // nested array with invalid type
        final var schema = NestedItems.NestedItems1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    Arrays.asList(
                        Arrays.asList(
                            Arrays.asList(
                                "1"
                            )
                        ),
                        Arrays.asList(
                            Arrays.asList(
                                2
                            ),
                            Arrays.asList(
                                3
                            )
                        )
                    ),
                    Arrays.asList(
                        Arrays.asList(
                            Arrays.asList(
                                4
                            ),
                            Arrays.asList(
                                5
                            ),
                            Arrays.asList(
                                6
                            )
                        )
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNotDeepEnoughFails() {
        // not deep enough
        final var schema = NestedItems.NestedItems1.getInstance();
        try {
            schema.validate(
                Arrays.asList(
                    Arrays.asList(
                        Arrays.asList(
                            1
                        ),
                        Arrays.asList(
                            2
                        ),
                        Arrays.asList(
                            3
                        )
                    ),
                    Arrays.asList(
                        Arrays.asList(
                            4
                        ),
                        Arrays.asList(
                            5
                        ),
                        Arrays.asList(
                            6
                        )
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testValidNestedArrayPasses() throws ValidationException {
        // valid nested array
        final var schema = NestedItems.NestedItems1.getInstance();
        schema.validate(
            new NestedItems.NestedItemsListBuilder()
                .add(
                    Arrays.asList(
                        Arrays.asList(
                            Arrays.asList(
                                1
                            )
                        ),
                        Arrays.asList(
                            Arrays.asList(
                                2
                            ),
                            Arrays.asList(
                                3
                            )
                        )
                    )
                )
                .add(
                    Arrays.asList(
                        Arrays.asList(
                            Arrays.asList(
                                4
                            ),
                            Arrays.asList(
                                5
                            ),
                            Arrays.asList(
                                6
                            )
                        )
                    )
                )
            .build(),
            configuration
        );
    }
}

File: samples/client/3_1_0_unit_test/kotlin/src/test/kotlin/org/openapijsonschematools/client/components/schemas/NestedItemsTest.kt
package org.openapijsonschematools.client.components.schemas

import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags
import org.openapijsonschematools.client.configurations.SchemaConfiguration
import org.openapijsonschematools.client.exceptions.ValidationException

import kotlin.test.Test
import kotlin.test.assertFailsWith

class NestedItemsTest {
    companion object {
        val configuration = SchemaConfiguration(JsonSchemaKeywordFlags.Builder().format().build())
    }

    @Test
    fun testNestedArrayWithInvalidTypeFails() {
        // nested array with invalid type
        val schema = NestedItems.NestedItems1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        listOf(
                            listOf(
                                listOf(
                                    "1"
                                )
                            ),
                            listOf(
                                listOf(
                                    2
                                ),
                                listOf(
                                    3
                                )
                            )
                        ),
                        listOf(
                            listOf(
                                listOf(
                                    4
                                ),
                                listOf(
                                    5
                                ),
                                listOf(
                                    6
                                )
                            )
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNotDeepEnoughFails() {
        // not deep enough
        val schema = NestedItems.NestedItems1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    listOf(
                        listOf(
                            listOf(
                                1
                            ),
                            listOf(
                                2
                            ),
                            listOf(
                                3
                            )
                        ),
                        listOf(
                            listOf(
                                4
                            ),
                            listOf(
                                5
                            ),
                            listOf(
                                6
                            )
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testValidNestedArrayPasses() {
        // valid nested array
        val schema = NestedItems.NestedItems1.getInstance()
        schema.validate(
            NestedItems.NestedItemsListBuilder()
                .add(
                    listOf(
                        listOf(
                            listOf(
                                1
                            )
                        ),
                        listOf(
                            listOf(
                                2
                            ),
                            listOf(
                                3
                            )
                        )
                    )
                )
                .add(
                    listOf(
                        listOf(
                            listOf(
                                4
                            ),
                            listOf(
                                5
                            ),
                            listOf(
                                6
                            )
                        )
                    )
                )
            .build(),
            configuration
        )
    }
}
